tiledb/sm/array_schema/enumeration.cc

```diff
@@ -295,47 +296,61 @@ shared_ptr<const Enumeration> Enumeration::extend(
       throw EnumerationException(
           "Offsets size is non-zero when extending a fixed sized enumeration.");
     }
-  }
-
-  Buffer new_data(data_.size() + data_size);
-  throw_if_not_ok(new_data.write(data_.data(), data_.size()));
-  if (data_size > 0) {
-    throw_if_not_ok(new_data.write(data, data_size));
-  }
 
-  const void* new_offsets_ptr = nullptr;
-  uint64_t new_offsets_size = 0;
-
-  Buffer new_offsets(offsets_.size() + offsets_size);
-
-  if (var_size()) {
-    // First we write our existing offsets
-    throw_if_not_ok(new_offsets.write(offsets_.data(), offsets_.size()));
-
-    // All new offsets have to be rewritten to be relative to the length
-    // of the current data array.
-    const uint64_t* offsets_arr = static_cast<const uint64_t*>(offsets);
-    uint64_t num_offsets = offsets_size / sizeof(uint64_t);
-    for (uint64_t i = 0; i < num_offsets; i++) {
-      uint64_t new_offset = offsets_arr[i] + data_.size();
-      throw_if_not_ok(new_offsets.write(&new_offset, sizeof(uint64_t)));
+    if (data_size % cell_size() != 0) {
+      throw EnumerationException(
+          "Invalid data size is not a multiple of the cell size.");
     }
-
-    new_offsets_ptr = new_offsets.data();
-    new_offsets_size = new_offsets.size();
   }
 
-  return create(
+  // Construct an empty enumeration to merge the old and new data
+  auto extended_enumeration = std::const_pointer_cast<Enumeration>(create(
       name_,
       "",
       type_,
       cell_val_num_,
       ordered_,
-      new_data.data(),
-      new_data.size(),
-      new_offsets_ptr,
-      new_offsets_size,
```

Comment:
We should add a constructor overload that takes two `Buffer&&`s for the data and the offsets. This way we maintain C.41.